metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlTable.java

```diff
@@ -0,0 +1,234 @@
+            uris.forEach(uri -> {
+                queryBuilder.append(" or location like ?");
+                params.add(new SqlParameterValue(Types.VARCHAR, uri + "%"));
+            });
+            queryBuilder.append(" )");
+        } else {
+            queryBuilder.append(" and location in (");
+            uris.forEach(uri -> {
+                queryBuilder.append("?,");
+                params.add(new SqlParameterValue(Types.VARCHAR, uri));
+            });
+            queryBuilder.deleteCharAt(queryBuilder.length() - 1).append(")");
+        }
+        ResultSetExtractor<Map<String, List<QualifiedName>>> handler = rs -> {
+            final Map<String, List<QualifiedName>> result = Maps.newHashMap();
+            while (rs.next()) {
+                final String schemaName = rs.getString("schema_name");
+                final String tableName = rs.getString("table_name");
+                final String uri = rs.getString("location");
+                final List<QualifiedName> names = result.computeIfAbsent(uri, k -> Lists.newArrayList());
+                names.add(QualifiedName.ofTable(catalogName, schemaName, tableName));
+            }
+            return result;
+        };
+        try {
+            return jdbcTemplate.query(queryBuilder.toString(), params.toArray(), handler);
+        } finally {
+            this.fastServiceMetric.recordTimer(
+                HiveMetrics.TagGetTableNames.getMetricName(), registry.clock().wallTime() - start);
+        }
+    }
+
+    /**
+     * Locks the iceberg table for update so that no other request can modify the table.
+     * @param tableId table internal id
+     * @param tableName table name
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void lockIcebergTable(final Long tableId, final QualifiedName tableName) {
+        String tableType = null;
+        try {
+            tableType = jdbcTemplate.queryForObject(SQL.TABLE_PARAM_LOCK,
+                new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
+                    new SqlParameterValue(Types.VARCHAR, PARAM_TABLE_TYPE), }, String.class);
+        } catch (EmptyResultDataAccessException ignored) { }
+        if (tableType == null || !ICEBERG_TABLE_TYPE.equals(tableType)) {
+            throw new InvalidMetaException(tableName, "Originally table is not of type iceberg", null);
+        }
+        Boolean isLocked = null;
+        try {
+            isLocked = jdbcTemplate.queryForObject(SQL.TABLE_PARAM_LOCK,
+                new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
+                    new SqlParameterValue(Types.VARCHAR, PARAM_METADATA_LOCK), }, Boolean.class);
+        } catch (EmptyResultDataAccessException ignored) { }
+        if (isLocked == null) {
+            insertTableParam(tableId, PARAM_METADATA_LOCK, 1);
+        } else if (isLocked) {
+            throw new IllegalStateException("Iceberg table is locked.");
+        } else {
```

Comment:
We can include the tableName in the exception string, which can help us to trace later. I am not sure if this message will be printed out somewhere. In case there're multiple locking failures from our side, we need a way to figure out which table has locking problems. We can do atlas metrics too, but think logging it out is easier here.